File: go/tree/101_is_symmetric.go
package tree

// isSymmetric 对称二叉树 BFS 实现
func isSymmetric(root *TreeNode) bool {

	if root == nil {
		return true
	}

	// 遍历每一层的数据，查看是否相同
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) > 0 {

		length := len(queue)
		// 此处判断对称逻辑不完整, 
        // 加上node.Left == nil 时append nil, 同时判断queue[i] 是否为nil
		for i, j := 0, length-1; i < j; i, j = i+1, j-1 {
            if queue[i] == nil && queue[j] == nil {
                continue
            }
            if queue[i] == nil || queue[j] == nil {
                return false
            }

			if queue[i].Val != queue[j].Val {
				return false
			}
		}

		for i := 0; i < length; i++ {
			node := queue[0]
			queue = queue[1:]

			if node == nil {
				continue
			}

			if node.Left != nil {
				queue = append(queue, node.Left)
			} else {
				queue = append(queue, nil)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			} else {
				queue = append(queue, nil)
			}
		}

	}
	return true
}

// isSymmetric 对称二叉树 DFS 实现
func isSymmetric2(root *TreeNode) bool {

	if root == nil {
		return true
	}

	var check func(*TreeNode, *TreeNode) bool

	check = func(left, right *TreeNode) bool {
		// 都为nil
		if left == nil && right == nil {
			return true
		}
		// 一个为nil 另一个补位nil false
		if left == nil || right == nil {
			return false
		}

		// 值相同，然后递归比较左右子树 是否对称
		return left.Val == right.Val && check(left.Left, right.Right) &&
			check(left.Right, right.Left)
	}
	return check(root.Left, root.Right)
}

File: rust/src/tree/_101_is_symmetric.rs
use super::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

// is_symmetric 对称二叉树 BFS 实现
pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
    if root.is_none() {
        return true;
    }

    use std::collections::VecDeque;
    let mut queue = VecDeque::new();

    // 把root 的左右子树放到queue 中，查看左右子树是否对称(相邻的左右子树应该对称)
    if let Some(root) = root {
        queue.push_back(root.borrow_mut().left.take());
        queue.push_back(root.borrow_mut().right.take());

        // queue 中的数据出队列，查看是否相同
        while let (Some(n1), Some(n2)) = (queue.pop_front(), queue.pop_front()) {
            match (n1, n2) {
                (None, None) => (), // 都是None  跳过
                (Some(x), Some(y)) => {
                    if x.borrow().val == y.borrow().val {
                        queue.push_back(x.borrow_mut().left.take());
                        queue.push_back(y.borrow_mut().right.take());
                        queue.push_back(x.borrow_mut().right.take());
                        queue.push_back(y.borrow_mut().left.take());
                    } else {
                        return false;
                    }
                }
                _ => return false,
            }
        }
    }

    true
}

// is_symmetric 对称二叉树 DFS 实现
pub fn is_symmetric2(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
    if root.is_none() {
        return true;
    }

    if let Some(n) = root {
        return check(&n.borrow().left, &n.borrow().right);
    }
    true
}

fn check(left: &Option<Rc<RefCell<TreeNode>>>, right: &Option<Rc<RefCell<TreeNode>>>) -> bool {
    if left.is_none() && right.is_none() {
        return true;
    }
    // left, right 其中一个为None 另一个不为None
    if left.is_none() || right.is_none() {
        return false;
    }

    if let (Some(l), Some(r)) = (left, right) {
        return l.borrow().val == r.borrow().val
            && check(&l.borrow().left, &r.borrow().right)
            && check(&l.borrow().right, &r.borrow().left);
    }
    true
}

// is_symmetric 对称二叉树 DFS 实现
pub fn is_symmetric3(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
    if root.is_none() {
        return true;
    }

    if let Some(root) = root {
        return check_val(root.borrow().left.clone(), root.borrow().right.clone());
    }
    false
}

fn check_val(left: Option<Rc<RefCell<TreeNode>>>, right: Option<Rc<RefCell<TreeNode>>>) -> bool {
    match (left, right) {
        (Some(l), Some(r)) => {
            l.borrow().val == r.borrow().val
                && check_val(l.borrow().left.clone(), l.borrow().right.clone())
                && check_val(l.borrow().right.clone(), l.borrow().left.clone())
        }

        (None, None) => true,
        _ => false,
    }
}
